faunadb-java/src/main/java/com/faunadb/client/query/Pagination.java

```diff
@@ -0,0 +1,164 @@
+package com.faunadb.client.query;
+
+import com.faunadb.client.types.Value;
+import com.google.common.collect.ImmutableMap;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A pagination expression.
+ * <p>
+ * <b>Example:</b>
+ * We can paginate all instances of a class, with its evets, in pages with 4 elements only.
+ * <pre>{@code
+ * client.query(
+ *    Paginate(Match(Ref("indexes/all_spells"))
+ *      .withEvents(true)
+ *      .withSize(4)
+ * );
+ * }</pre>
+ *
+ * @see Language#Paginate(Value)
+ * @see <a href="https://faunadb.com/documentation/queries#read_functions">FaunaDB Read Functions</a>
+ */
+public final class Pagination extends Fn {
+
+  /**
+   * Helper to construct the pagination cursor that can constructed either
+   * by {@link Language#Before(Value)} or {@link Language#After(Value)}
+   *
+   * @see <a href="https://faunadb.com/documentation/queries#read_functions">FaunaDB Read Functions</a>
+   */
+  public abstract static class Cursor {
+    private final String name;
+    private final Value ref;
+
+    Cursor(String name, Value ref) {
+      this.name = requireNonNull(name);
+      this.ref = requireNonNull(ref);
+    }
+  }
+
+  static class Before extends Cursor {
+    Before(Value ref) {
+      super("before", ref);
```

Comment:
rather than storing the name as a string, you could do a class check when building the call instead.